File: packages/shared/src/referral-share.test.ts
import { describe, expect, it } from "vitest";

import { REFERRAL_REWARDS_PER_YEAR } from "./referrals";
import {
  REFERRAL_ASK_MIN_CUSTOMERS,
  REFERRAL_ASK_MIN_DAYS,
  REFERRAL_ASK_QUIET_DAYS,
  REFERRAL_ASK_ACTION,
  REFERRAL_ASK_BODY,
  REFERRAL_ASK_DISMISS,
  REFERRAL_REWARD_LINE,
  REFERRAL_SHARE_ACTION,
  REFERRAL_SHARE_COPIED,
  REFERRAL_SHARE_COPY,
  REFERRAL_SHARE_DRAFT_LABEL,
  REFERRAL_SHARE_LINK_NOTE,
  REFERRAL_SHARE_NOTE,
  REFERRAL_SHARE_TITLE,
  REFERRAL_STAGE_LABELS,
  referralAskDecision,
  referralAskHeadline,
  referralShareText,
  type ReferralAskFacts,
} from "./referral-share";

import { EN as WEB_EN, FR_CA as WEB_FR } from "../../../apps/web/src/i18n/catalog";

/* #228 — this module names keys, so the assertions resolve them. */
function lookUp(table: unknown, key: string, lang: string): string {
  const [section, name] = key.split(".");
  const value = (table as Record<string, Record<string, string>>)[section]?.[name];
  if (typeof value !== "string") throw new Error(`no ${lang} for ${key}`);
  return value;
}

const say = (key: string): string => lookUp(WEB_EN, key, "English");

/* The share fallback carries {code}, so its resolver substitutes. */
function withVars(table: unknown, lang: string) {
  return (key: string, vars: Record<string, string>): string =>
    Object.entries(vars).reduce(
      (out, [token, value]) => out.split(`{${token}}`).join(value),
      lookUp(table, key, lang),
    );
}

const sayVars = withVars(WEB_EN, "English");
const sayVarsFr = withVars(WEB_FR, "French");
const sayFr = (key: string): string => lookUp(WEB_FR, key, "French");

const DAY = 24 * 60 * 60 * 1000;
const NOW = new Date("2026-08-08T12:00:00.000Z");

function ago(days: number): string {
  return new Date(NOW.getTime() - days * DAY).toISOString();
}

/** A workspace that has earned the ask, so each test can break one thing. */
function earned(overrides: Partial<ReferralAskFacts> = {}): ReferralAskFacts {
  return {
    activated: true,
    activatedAt: ago(60),
    repliedCustomers: REFERRAL_ASK_MIN_CUSTOMERS + 5,
    dismissedAt: null,
    rewardsThisYear: 0,
    rewardCap: REFERRAL_REWARDS_PER_YEAR,
    ...overrides,
  };
}

describe("referralShareText", () => {
  it("puts the link on the end so an edited message keeps it", () => {
    const text = referralShareText("Come try this", "https://loonext.com/?ref=ABCD2345", "ABCD2345", sayVars);
    expect(text).toBe("Come try this\n\nhttps://loonext.com/?ref=ABCD2345");
  });

  it("survives the owner deleting every word of it", () => {
    // The whole reason the link is not inside the editable box. An empty draft
    // must still send something usable rather than a lone blank line.
    expect(referralShareText("   ", "https://loonext.com/?ref=ABCD2345", "ABCD2345", sayVars)).toBe(
      "https://loonext.com/?ref=ABCD2345",
    );
  });

  it("falls back to the code when there is no link to give", () => {
    // SITE_ORIGIN unset. A code read aloud at a supply counter is how a fair
    // share of these will travel anyway; "undefined/?ref=" is not.
    expect(referralShareText("Have a look", null, "ABCD2345", sayVars)).toBe(
      "Have a look\n\nUse my code ABCD2345 when you sign up.",
    );
  });

  it("#228 gives a French owner a French line under their own words", () => {
    // Unlike the automated bodies in locale.ts, this is not sent by us. It is
    // dropped into a share sheet for the OWNER to send, sitting under words
    // they typed - so one English line at the bottom of a French message is
    // the product speaking out of turn in somebody else's conversation.
    const text = referralShareText("Essayez ceci", null, "ABCD2345", sayVarsFr);
    expect(text).toContain("ABCD2345");
    expect(text).toContain("Utilisez mon code");
    expect(text).not.toContain("when you sign up");
    expect(text).not.toMatch(/\{/);
  });

  it("#228 never reaches the fallback sentence when a link exists", () => {
    // The link case is the common one and carries no sentence at all, in
    // either language - a URL is not copy.
    for (const resolve of [sayVars, sayVarsFr]) {
      const text = referralShareText("x", "https://x.test/?ref=A", "A", resolve);
      expect(text).toBe("x\n\nhttps://x.test/?ref=A");
    }
  });

  it("trims the draft rather than sending trailing whitespace", () => {
    expect(referralShareText("Look  \n\n", "https://x.test/?ref=A", "A", sayVars)).toBe(
      "Look\n\nhttps://x.test/?ref=A",
    );
  });
});

describe("REFERRAL_SHARE_NOTE", () => {
  it("does not claim the crew is limited to one number", () => {
    // A crew can run several numbers, so "one number" is a promise the product
    // does not make. The inbox is the thing that is singular.
    expect(REFERRAL_SHARE_NOTE).not.toMatch(/one number/i);
  });

  it("carries no link of its own", () => {
    // referralShareText appends it. A URL in the default draft would be a
    // second place for the link to come from, and one of the two would be wrong.
    expect(REFERRAL_SHARE_NOTE).not.toMatch(/https?:/);
  });

  it("fits in a text message somebody will actually read", () => {
    expect(REFERRAL_SHARE_NOTE.length).toBeLessThanOrEqual(220);
  });
});

describe("referralAskDecision", () => {
  it("asks a workspace that has been working for a month", () => {
    expect(referralAskDecision(earned(), NOW)).toEqual({
      ask: true,
      customers: REFERRAL_ASK_MIN_CUSTOMERS + 5,
    });
  });

  it("never asks at signup", () => {
    // The acceptance criterion, stated as a test: a workspace the product has
    // not worked for once is not asked to vouch for it.
    expect(
      referralAskDecision(earned({ activated: false, activatedAt: null }), NOW),
    ).toEqual({ ask: false, refusal: "not_activated" });
  });

  it("does not ask a workspace that activated yesterday", () => {
    expect(referralAskDecision(earned({ activatedAt: ago(1) }), NOW).refusal).toBe(
      "too_new",
    );
  });

  it("counts the month from the day it started working, not from signup", () => {
    // One day short, then one day over. The boundary is the whole point of the
    // constant, so it is asserted rather than assumed.
    expect(
      referralAskDecision(
        earned({ activatedAt: ago(REFERRAL_ASK_MIN_DAYS - 1) }),
        NOW,
      ).ask,
    ).toBe(false);
    expect(
      referralAskDecision(earned({ activatedAt: ago(REFERRAL_ASK_MIN_DAYS) }), NOW)
        .ask,
    ).toBe(true);
  });

  it("does not ask a crew the product is barely doing anything for", () => {
    expect(
      referralAskDecision(
        earned({ repliedCustomers: REFERRAL_ASK_MIN_CUSTOMERS - 1 }),
        NOW,
      ),
    ).toEqual({ ask: false, refusal: "too_quiet" });
  });

  it("takes 'Not now' as an answer for a quarter", () => {
    expect(
      referralAskDecision(earned({ dismissedAt: ago(1) }), NOW).refusal,
    ).toBe("dismissed");
    expect(
      referralAskDecision(
        earned({ dismissedAt: ago(REFERRAL_ASK_QUIET_DAYS - 1) }),
        NOW,
      ).ask,
    ).toBe(false);
    expect(
      referralAskDecision(
        earned({ dismissedAt: ago(REFERRAL_ASK_QUIET_DAYS) }),
        NOW,
      ).ask,
    ).toBe(true);
  });

  it("stops asking once the reward can no longer be paid", () => {
    // Not a timing rule — an honesty one. decideReferral refuses a capped
    // referrer's claim, so asking anyway would offer a month we have already
    // decided not to pay, and their friend is the one who finds out.
    expect(
      referralAskDecision(
        earned({ rewardsThisYear: REFERRAL_REWARDS_PER_YEAR }),
        NOW,
      ),
    ).toEqual({ ask: false, refusal: "capped" });
    expect(
      referralAskDecision(
        earned({ rewardsThisYear: REFERRAL_REWARDS_PER_YEAR - 1 }),
        NOW,
      ).ask,
    ).toBe(true);
  });

  it("refuses rather than guesses when a timestamp is unreadable", () => {
    expect(referralAskDecision(earned({ activatedAt: "not a date" }), NOW).ask).toBe(
      false,
    );
    // A dismissal we cannot parse is still a dismissal. Somebody pressed the
    // button; our parsing problem must not turn that into a yes.
    expect(
      referralAskDecision(earned({ dismissedAt: "not a date" }), NOW),
    ).toEqual({ ask: false, refusal: "dismissed" });
  });

  it("treats activated-without-a-timestamp as not activated", () => {
    // The two halves come from different columns, and a row where one is set
    // and the other is not would otherwise fall through to a NaN comparison.
    expect(
      referralAskDecision(earned({ activatedAt: null }), NOW).refusal,
    ).toBe("not_activated");
  });
});

describe("referralAskHeadline", () => {
  it("counts in customers, in their own numbers", () => {
    expect(referralAskHeadline(37, say)).toBe("You replied to 37 customers this month.");
  });

  it("does not say '1 customers'", () => {
    expect(referralAskHeadline(1, say)).toBe("You replied to 1 customer this month.");
  });

  it("does not say it in French either", () => {
    // The reason singular and plural are separate keys rather than one sentence
    // with the number swapped in. A French translation that reused the plural
    // would read "1 clients", which is the exact defect the English test above
    // exists to prevent, in the language nobody re-reads.
    expect(referralAskHeadline(1, sayFr)).not.toMatch(/1 clients/);
    expect(referralAskHeadline(37, sayFr)).toContain("37");
    expect(referralAskHeadline(1, sayFr)).not.toBe(referralAskHeadline(37, sayFr));
  });
});

/*
 * #228 — every key this module names, resolved in both languages.
 *
 * Written up front rather than after a break-test found the gap: these are
 * fifteen exported constants that clients render directly, so a typo in any one
 * reaches a screen as its own name and nothing else here would notice.
 */
describe("#228 every referral string exists in both languages", () => {
  it("resolves all of them", () => {
    const keys = [
      REFERRAL_SHARE_NOTE,
      REFERRAL_SHARE_TITLE,
      REFERRAL_REWARD_LINE,
      REFERRAL_SHARE_ACTION,
      REFERRAL_SHARE_COPY,
      REFERRAL_SHARE_COPIED,
      REFERRAL_SHARE_DRAFT_LABEL,
      REFERRAL_SHARE_LINK_NOTE,
      REFERRAL_ASK_BODY,
      REFERRAL_ASK_ACTION,
      REFERRAL_ASK_DISMISS,
      ...Object.values(REFERRAL_STAGE_LABELS),
    ];
    for (const key of keys) {
      expect(say(key).length, key).toBeGreaterThan(0);
      expect(sayFr(key).length, key).toBeGreaterThan(0);
      expect(sayFr(key), `${key} is not translated`).not.toBe(say(key));
    }
    expect(keys.length).toBeGreaterThan(14);
  });

  it("gives each referral stage its own words", () => {
    // Five stages sharing a label would satisfy every assertion above while
    // making the progress list say nothing.
    const labels = Object.values(REFERRAL_STAGE_LABELS).map(say);
    expect(new Set(labels).size).toBe(Object.keys(REFERRAL_STAGE_LABELS).length);
  });
});

File: apps/android/app/src/test/kotlin/com/loonext/android/core/referral/ReferralShareTest.kt
package com.loonext.android.core.referral

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #288 — the share draft, and that this phone offers the same words the laptop does.
 *
 * Two halves. The behaviour tests assert the arithmetic and the one rule that
 * matters (an edited message keeps its link); the parity tests read
 * `packages/shared/src/referral-share.ts`, because this is a hand-port and
 * nothing about Kotlin says the original stayed put.
 *
 * A draft that reads differently on the phone than on the laptop is a draft an
 * owner stops trusting to say what they meant — and this is the one message in
 * the product a customer sends to somebody who has never heard of us.
 */
class ReferralShareTest {

    @Test
    fun `the link goes on the end, so an edited message keeps it`() {
        assertEquals(
            "Come try this\n\nhttps://loonext.com/?ref=ABCD2345",
            ReferralShare.shareText("Come try this", "https://loonext.com/?ref=ABCD2345", "ABCD2345"),
        )
    }

    @Test
    fun `an owner who deletes every word still sends something usable`() {
        // The whole reason the link is not inside the editable field.
        assertEquals(
            "https://loonext.com/?ref=ABCD2345",
            ReferralShare.shareText("   ", "https://loonext.com/?ref=ABCD2345", "ABCD2345"),
        )
    }

    @Test
    fun `with no link configured the code carries the referral`() {
        assertEquals(
            "Have a look\n\nUse my code ABCD2345 when you sign up.",
            ReferralShare.shareText("Have a look", null, "ABCD2345"),
        )
    }

    @Test
    fun `the draft trims rather than sending trailing whitespace`() {
        assertEquals(
            "Look\n\nhttps://x.test/?ref=A",
            ReferralShare.shareText("Look  \n\n", "https://x.test/?ref=A", "A"),
        )
    }

    @Test
    fun `the headline does not say 1 customers`() {
        assertEquals("You replied to 1 customer this month.", ReferralShare.askHeadline(1))
        assertEquals("You replied to 37 customers this month.", ReferralShare.askHeadline(37))
    }

    @Test
    fun `a stage this build does not know reads as itself rather than crashing`() {
        // A server ahead of this app. One unfamiliar row on a settings card is
        // recoverable; a crash where the card used to be is not.
        assertEquals("Up and running", ReferralShare.stageLabel("signed_up"))
        assertEquals("kaleidoscope", ReferralShare.stageLabel("kaleidoscope"))
    }

    @Test
    fun `the default draft promises nothing the product does not do`() {
        // A crew can run several numbers, so "one number" is not ours to claim.
        assertFalse(ReferralShare.NOTE.contains("one number", ignoreCase = true))
        // And it carries no link of its own: shareText appends it, and a URL in
        // here would be a second place for the link to come from.
        assertFalse(ReferralShare.NOTE.contains("http"))
    }

    // ---------------------------------------------------- against the original

    private fun repoFile(relative: String): String {
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(dir, relative)
            if (candidate.exists()) return candidate.readText()
            dir = dir.parentFile
        }
        throw AssertionError("$relative not found from ${File("").absolutePath}")
    }

    private fun shared(): String = repoFile("packages/shared/src/referral-share.ts")

    /**
     * The web catalogue's ENGLISH half, which is where the shared module's
     * sentences went.
     *
     * #228 moved `referral-share.ts` from holding sentences to naming keys, so
     * a `contains` against that file asks whether it holds a paragraph it no
     * longer holds. The guard's job is unchanged — this client must not drift
     * from the shared vocabulary — so it follows the words rather than being
     * deleted.
     *
     * Sliced to the English half: the French holds the same keys, and a
     * `contains` over the whole file would be asking whether a sentence
     * appears in EITHER language.
     */
    private fun catalogue(): String =
        repoFile("apps/web/src/i18n/sections/domain.ts")
            .substringAfter("export const domainEn")
            .substringBefore("export const domainFr")

    @Test
    fun `every sentence a crew might send matches the shared module`() {
        val source = catalogue()
        for (copy in listOf(
            ReferralShare.NOTE,
            ReferralShare.TITLE,
            ReferralShare.ACTION,
            ReferralShare.COPY,
            ReferralShare.COPIED,
            ReferralShare.DRAFT_LABEL,
            ReferralShare.LINK_NOTE,
            ReferralShare.REWARD_LINE,
            ReferralShare.ASK_BODY,
            ReferralShare.ASK_ACTION,
            ReferralShare.ASK_DISMISS,
        )) {
            // Long strings wrap across lines with `+` in both files, so the
            // comparison is against the source with those joins collapsed
            // rather than against the raw text.
            assertTrue(
                "this copy has drifted from the shared module: $copy",
                joined(source).contains(copy),
            )
        }
    }

    @Test
    fun `the stage labels match the shared module, all five`() {
        /*
         * #228: the shared module maps each stage to a KEY, and the catalogue
         * says what the key means. Pinning the whole chain rather than half of
         * it — a stage pointed at the wrong key and a key with the wrong words
         * are different bugs and this catches both.
         */
        val declared = Regex("""REFERRAL_STAGE_LABELS[^=]*= \{([^}]+)\}""")
            .find(joined(shared()))
            ?.groupValues
            ?.get(1)
            ?: throw AssertionError("REFERRAL_STAGE_LABELS is no longer an object literal")
        val pairs = Regex("""(\w+): "(domain\.\w+)"""").findAll(declared)
            .associate { it.groupValues[1] to it.groupValues[2] }
        assertEquals(5, pairs.size)

        val words = catalogue()
        for ((stage, key) in pairs) {
            val label = ReferralShare.stageLabel(stage)
            assertTrue(
                "the label for '$stage' has drifted from the catalogue: `$label`",
                joined(words).contains(label),
            )
            assertTrue(
                "'$stage' names $key, which the catalogue does not answer",
                words.contains(key.removePrefix("domain.") + ":"),
            )
        }
    }

    @Test
    fun `the fallback sentence for a missing link matches too`() {
        // Built by interpolation on both sides, so it is compared as its shape
        // rather than as a whole string.
        assertTrue(joined(shared()).contains("Use my code \${code} when you sign up."))
    }

    /**
     * The TypeScript source with its multi-line string concatenations joined.
     *
     * `"a " +\n  "b"` in the file is the single string `"a b"` at runtime, and a
     * literal `contains` against the raw text would fail on every sentence long
     * enough to wrap — which is most of them here.
     */
    private fun joined(source: String): String =
        source.replace(Regex(""""\s*\+\s*\n\s*""""), "")
}
